File: 0242-valid-anagram/solution.java
class Solution {
    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length())
            return false;
        
        HashMap<Character, Integer> sCharCount = new HashMap<>();

        for(char c : s.toCharArray()) {
            if(sCharCount.containsKey(c))
                sCharCount.put(c, sCharCount.get(c) + 1);
            else
                sCharCount.put(c, 1);
        }

        for(char c : t.toCharArray()) {
            if(sCharCount.containsKey(c)) {
                sCharCount.put(c, sCharCount.get(c) - 1);
                if(sCharCount.get(c) < 0)
                    return false;
            }
            else
                return false;
        }

        return true;
    }
}

File: 0242-valid-anagram/solution.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        s_map: dict[str, int] = {}
        for c in s:
            s_map[c] = s_map.get(c, 0) + 1
        for c in t:
            if c in s_map:
                s_map[c] -= 1
            else:
                return False

        for v in s_map.values():
            if v != 0:
                return False

        return True
        
